### DISCUSSION TITLE
Tile layout clarification

### DISCUSSION THREAD

**Discussion by trus-uic** [rajewsky-lab/openst#24] [Q&A] [answered]
Hello,

Thank you for providing and documenting this innovative tool, I am very excited to put it to use!

I wanted to reach out with a couple questions on the tiling structure within each lane. When examining Fig. S1, panel A of your preprint: the tile layout of a capture area is depicted as having an alternating y-offset for each tile within a row. This is the case with the fc_1 coordinate system depicting the tiles having alternating offsets in-row.
![image](https://github.com/rajewsky-lab/openst/assets/142931687/9083a61b-136e-4fbb-8c42-ee87cedeb14c)

In Illumina's depiction of their tiling structure for an S4, their tiles are in a grid-format:
![image](https://github.com/rajewsky-lab/openst/assets/142931687/55ca8ae3-d924-408c-9eaf-8d10625f7a63)
https://emea.illumina.com/content/dam/illumina/gcs/assembled-assets/marketing-literature/novaseq-lane-edge-tech-note-m-amr-00104/novaseq-lane-edge-tech-note-m-amr-00104.pdf - page 3

Could you please clarify why this is the case? Is this the actual tiling structure as defined by Illumina, which they simplify in the above representation?

Additionally, the fc coordinate system file depicts their being a column (swath) 0 within a lane (tiles x0xx), while your pre-print and Illumina's documentation define 6 swaths within a lane. Is this something that should be considered, or should lane 0 be considered as a result of the code index starting at 0?

Finally, the `fc_1_coordinate_system.csv` provided for the e13 mouse head sample and `fc_2_coordinate_system.csv` for adult mouse hippocampus sample are incredibly similar, except that they start at different initial offsets for the leftmost swath within each lane. Could you clarify how this offset is determined or in general how the lane-tile coordinate system is determined?

Thank you for your time and please let me know if I am misinterpreting anything above. Much appreciated.

edit: grammar

**Comment by nukappa** [comment, marked as answer]
Hi @trus-uic, thanks for the interest!

1. Indeed, the scheme provided by Illumina is simplified.
2. Nomenclature is the following. For `fc_1_L3_tile_1056`, for instance, `L3` denotes the lane number, `56` is the tile number within the same swath, `0` denotes the swath number, and `1` in the `1056` means the bottom part of the flow cell was used (`2` if the top).
3. We determined the lane-tile coordinate systems empirically, by stitching our data and also comparing to the H&E image. Seems that these two Illumina flow cells were not manufactured in the same way.

Hope this helps & best of luck with implementing open-ST!

**Reply by trus-uic** [reply]
Thank you for your clarification! Best of luck.

**Comment by gianmore** [comment]
Hi! thanks for the amazing tool you developed! We have now generated according to your protocol pucks from an S4 flowcell. We have previously sequenced the capture area and we have retrieved the spatial barcodes coordinates for all tiles. So now I have my lanes_and_tiles.txt file, my collection of puck_barcode_files, and for the coordinate_system.csv I simply took the one you provide and I have changed the names in the puck_id column according to my nomeclature of the puck_barcode_file. But instead the offset have to be experimentally determined for each flowcell?? I see your explanation here (We determined the lane-tile coordinate systems empirically, by stitching our data and also comparing to the H&E image. Seems that these two Illumina flow cells were not manufactured in the same way.) Can you provide a documentation/tutorial on how to do it? Thanks in advance!

**Reply by StevenSong** [reply]
@gianmore I'm working with @trus-uic's group on implementing this protocol. The way we've confirmed we're using the right tile coordinate system is by inspecting the barcode x gene count matrix generated by the spacemake step (as described in the [openst documentation here](https://rajewsky-lab.github.io/openst/latest/computational/preprocessing_openst_library/)). That is, we plug in one version of the tile coordinate system to generate the stitched expression matrix and plot a heat map of barcodes with non-zero counts. This results in a plot like the following:

![all_genes](https://github.com/user-attachments/assets/3d52ae2b-8d4b-4086-9e56-cedad539a5d0)

Even without seeing the H&E, you can hopefully appreciate the edges of the tissue section are contiguous as they transition between lanes. There are definitely other artifacts in the plot but if we had used the wrong tile coordinate file, the edges would appear sheared. To help reason what that might look like, here's a visualization of what the 2 coordinate files look like relative to each other (I forget which is which).

![Screenshot 2024-08-19 at 12 49 21 PM (1)](https://github.com/user-attachments/assets/0adb4f8b-bfa8-474b-b3c0-61bdb6e21d61)

![Screenshot 2024-08-19 at 12 49 26 PM (1)](https://github.com/user-attachments/assets/f11ae04a-7356-4102-a753-e57ab2bcda42)

If the tissue looks sheared, you can stitch using the other tile coordinate file by running the `spatial_stitch` utility provided by openst (it's buried in the documentation under a dropdown, pasted below)
```
openst spatial_stitch \
    --tiles \
        dge.fc_1_tile_2267.h5ad \
        dge.fc_1_tile_2268.h5ad \
    --tile-id \
        fc_1_tile_2267 \
        fc_1_tile_2268 \
    --tile-coordinates fc_1_coordinate_system.csv \
    --h5-out spatial_stitched_spots.h5ad
```

Note that the stitched barcode coordinates in the resulting anndata object are at `obsm/spatial`, the coordinates at `obs/x|y_pos` are the intra-tile coordinates that do not incorporate the offset. If neither of the provided coordinate files look correct, you can manually adjust the stitching using the information in `obs`, though we haven't had to do this yet.

We also reached out to Illumina to ask if they tracked which layout was used by each flow cell but, perhaps unsurprisingly, they were not keen on answering that question and stated that flow cells were identical. We haven't looked at enough flow cells to confirm either way, but we're fairly happy with the heat map method.

**Reply by StevenSong** [reply]
Here's some code to generate the heat map from the stitched h5ad:

```python
import anndata
import numpy as np
import matplotlib.pyplot as plt

stitched = "dge.all.polyA_adapter_trimmed.mm_included.spatial_beads_puck_collection.h5ad"

dat = anndata.read_h5ad(stitched)

x = dat.obsm["spatial"][:, 0]
y = dat.obsm["spatial"][:, 1]

counts = np.asarray(dat.X.sum(axis=1)).T[0]
mask = counts > 0
new_x = x[mask]
new_y = y[mask]
fig, ax = plt.subplots(figsize=(10, 10))
marker_size = 0.0001
ax.scatter(new_x, new_y, s=marker_size, marker='o', facecolors="tab:blue", edgecolors="tab:blue")
ax.set_title(f"Any gene, N barcodes: {len(new_x)}, marker size: {marker_size:0.4f}")
fig.savefig("all_genes.png", dpi=300)
```

**Reply by gianmore** [reply]
Hi Steven, thank you very much! Just to check if I got it right, so if from the heatmap I can see that the borders between the tiles are sheared I can run openst spatial_stitch with the alternative coordinate system (it should work either fc_1_coordinate_system.csv or fc_2_coordinate_system.csv right?) and then I will have my final spatial_stitched_spots.h5ad file to continue the analysis with, is that correct?

**Reply by StevenSong** [reply]
yes, that's my understanding. you'd have to rename the tile IDs in the coordinate system csv's, like you mentioned in your original comment. hypothetically, if neither coordinate system csv's are correct, you'd have to manually offset the intra-tile coordinates stored in `obs/x|y_pos`, but we've had luck with `fc_1_coordinate_system.csv`

**Reply by gianmore** [reply]
Thanks, Steven! Do you happen to have a full image showing the numbering of the tiles in the lane? For the pucks we generated in-house, I'm considering dividing the puck barcode files into macroareas. This way, spacemake will have fewer files to screen for barcode matches.

**Reply by StevenSong** [reply]
Yes, I generated the following grids based on the coordinates in `fc_1_coordinate_system.csv` and `fc_2_coordinate_system.csv`. Note that these 2 files both have a 0 column (swath) per lane, which is not reflective of the actual flow cells which only have columns 1-6: see page 73 of the [support pdf](https://support.illumina.com/content/dam/illumina-support/documents/documentation/system_documentation/novaseq/1000000019358_16-novaseq-6000-system-guide.pdf). We asked the authors about this in the original body of this post but not sure if the question came across correctly. Anyways, we have 2 versions of either flow cell layout, one with and one without the 0 swath.

<details>
<summary>flowcell 1</summary>

![fc1](https://github.com/user-attachments/assets/f6a56dc0-fb5f-4fe9-93be-e70f0438b1c2)

</details>

<details>
<summary>flowcell 2</summary>

![fc2](https://github.com/user-attachments/assets/d33ffe06-30d4-4279-8f29-d1337aa0ec43)

</details>

<details>
<summary>flowcell 1 with 0 swath</summary>

![fc1col0](https://github.com/user-attachments/assets/6313119d-2ec3-4dcc-93eb-ce88d795a502)

</details>

<details>
<summary>flowcell 2 with 0 swath</summary>

![fc2col0](https://github.com/user-attachments/assets/389513a3-c7d9-4ec8-8f10-247935a7105c)

</details>

Last thing is about the macro-areas. We thought the same thing and simply grouped it by lane (1-4) and glass piece (1-2). So in the end we had 8 sets of tiles. That way, we didn't have to worry about column/row boundaries and which orientation the tiles go in, we weren't initially sure about these details (though after processing our first samples, I think we could do finer granularity). Still, maybe the lane/glass grouping is a nice middle ground

**Reply by gianmore** [reply]
Thank you very much! I share your concern about the coordinate system.
So in `fc_1_coordinate_system.csv` and `fc_2_coordinate_system.csv` there are basically 7 swath per lane included since the nomenclature goes from 0 to 6 and not 6 as indicated in the Illumina documentation (nomenclature from 1 to 6). My understanding is that the S4 have 6 swath per lane. @nukappa can you clarify this?
